File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap16/Serialization.pl
#!/usr/bin/perl -w
use strict;

###  Serialization.pl

use Storable;                                                       #(A)

my $pets = ["poodle", "chihuahua", "huskie"];                       #(B)
my $fruit = { banana  =>  { color =>  'yellow',                     #(C)
                            price =>  0.35 },
              orange  =>  { color => 'orange',
                            price =>  0.28 } };
my $current_year = 2008;                                            #(D)
my $name = "Beeblecrox";                                            #(E)

eval {                                                              #(F)
    store( [$pets, $fruit, $current_year, $name], 'test.txt' );     #(G)
};
if ($@) {                                                           #(H)
    print "Error in eval: $@";                                      #(I)
}

#----------------------------------------------------------------------
# Retrieving data from the disk file:

my $recover = retrieve( 'test.txt' );                               #(J)

print $recover, "\n";                # ARRAY(0x805191c)             #(K)
my $beasts = $recover->[0];                                         #(L)
print "@$beasts\n";                  # poodle chihuahua huskie      #(M)

my $froot = $recover->[1];                                          #(N)
print "$froot\n";                    # HASH(0x808b798)              #(O)
my @allkeys = keys( %$froot );                                      #(P)
print "@allkeys\n";                  # orange banana                #(Q)

print "$recover->[2]\n";             # 2008                         #(R)

print "$recover->[3]\n";             # Beeblecrox                   #(S)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap16/Serialization.py
#!/usr/bin/env python

###  Serialization.py

import pickle                                                       #(A)

pets  = ("poodle", "chihuahua", "huskie")                           #(B)
fruit = { "banana" : { "color" : 'yellow',                          #(C)
                       "price" :  0.35 },                           #(D)
          "orange" : { "color" : 'orange',                          #(E)
                       "price" :  0.28 } }                          #(F)
current_year = 2008                                                 #(G)
name = "Beeblecrox"                                                 #(H)

# Serialize the above objects into a single byte stream:
serialized_output = pickle.dumps((pets, fruit, current_year, name)) #(I)

# Recover the individual objects from the serialized stream:
recover = pickle.loads( serialized_output )                         #(J)

print recover[0]   # ('poodle', 'chihuahua', 'huskie')              #(K)

print recover[1]   # {'orange': {'color': 'orange',                 #(L)
                   #             'price': 0.28000000000000003},
                   #  'banana': {'color': 'yellow',
                   #                      'price': 0.34999999999999998}}
print recover[2]   # 2008                                           #(M)

print recover[3]   # Beeblecrox                                     #(N)

# Directly write out the serialization to a file:
pickle.dump((pets, fruit, current_year, name),open('backup.p', 'w'))#(O)

#-----------------------------------------------------------------------
# Restore the objects from the disk file:
restored = pickle.load( open( 'backup.p' ) )                        #(P)

print restored     # exactly the same as for the 'recover' string   #(Q)
